Backport thermal_zone_convert_outdoor_air_to_per_area from newer openstudio-standards

JanghyunJK opened the pull request:
Coming from,
- https://github.com/NREL/openstudio-bem-to-surrogate-gem/pull/202
- https://github.com/jmarrec/openstudio-standards/pull/27

I believe we can't use `OpenstudioStandards::ThermalZone` in our OS version 3.6.1 because it's a newer thing. I just noticed this while testing new measure. Need to open a companion PR on the 179D bem workflow side too.

* Companion PR: https://github.com/NREL/openstudio-bem-to-surrogate-gem/pull/280

Commit:
adding thermal_zone_convert_outdoor_air_to_per_area

jmarrec commented:
@JanghyunJK please note I rebased this branch onto the new develop, and force-pushed, make sure to update your local refs if you intend to make more changes.

jmarrec commented:
Note: this is essentially a backporrt of https://github.com/NREL/openstudio-standards/blob/412de97737369c3ee642237a83c8e5a6b1ab14be/lib/openstudio-standards/thermal_zone/thermal_zone.rb#L719-L780

But this was just MOVED, so you shouldn't have to do this. You should just use thermal_zone_convert_outdoor_air_to_per_area(thermal_zone).

https://github.com/NREL/openstudio-standards/commit/e1509d3e08299cf5788e35507314a576a9e5b979#diff-63ececc6d4a48507defb1fa8d804bca862c4289d844c7250187fe0e3ca1c51c6R558

 **Edit**: It actually doesn't exist either on this branch, nevermind

jmarrec commented:
@brianlball @JanghyunJK I realized something is missing in:
* #27 

These methods are NEVER registered. Because the file isn't loaded at all. THis was not updated to include it:

https://github.com/NREL/openstudio-standards/blob/424f60964def674ffd34b8442b74c505ece516a0/lib/openstudio-standards.rb#L244-L249

jmarrec committed:
Register all 179d rb files

jmarrec committed:
Add a test for DCV that exhibits the fact that the method doesn't exist

Minitest::UnexpectedError:         NameError: uninitialized constant OpenstudioStandards::ThermalZone

jmarrec committed:
Improve testing

jmarrec commented on lib/openstudio-standards.rb:
```diff
@@ -243,9 +243,16 @@ module OpenstudioStandards
   # 179D
   require_relative "#{stds}/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.Model"
+  require_relative "#{stds}/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.AirLoopHVAC"
+  require_relative "#{stds}/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.AirTerminalSingleDuctVAVReheat"
+  require_relative "#{stds}/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.CoilHeatingDXMultiSpeed"
+  require_relative "#{stds}/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.CoilHeatingDXSingleSpeed"
+  require_relative "#{stds}/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.FanVariableVolume"
+  require_relative "#{stds}/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.PlanarSurface"
+  require_relative "#{stds}/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.PlantLoop"
   require_relative "#{stds}/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.SpaceType"
   require_relative "#{stds}/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.SubSurface"
-  require_relative "#{stds}/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.PlantLoop"
+  require_relative "#{stds}/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.ThermalZone"
```
Register the files. (we glob on the openstudio-bem-to-surrogate-gem so we didn't have that problem)

Now we can test for them!

jmarrec commented on test/90_1_general/test_179d.rb:
```diff
@@ -511,4 +511,74 @@ def test_space_type_apply_internal_load_schedules_override
+    z.additionalProperties.setFeature("apxg no need to have DCV", true)
+    z.additionalProperties.setFeature("zone DCV implemented in user model", true)
+    refute(@standard.baseline_thermal_zone_demand_control_ventilation_required?(z))
+    refute(@standard.baseline_air_loop_hvac_demand_control_ventilation_required?(a))
+    z.additionalProperties.setFeature("apxg no need to have DCV", false)
+    z.additionalProperties.setFeature("zone DCV implemented in user model", false)
+    refute(@standard.baseline_thermal_zone_demand_control_ventilation_required?(z))
+    refute(@standard.baseline_air_loop_hvac_demand_control_ventilation_required?(a))
+    z.additionalProperties.setFeature("apxg no need to have DCV", false)
+    z.additionalProperties.setFeature("zone DCV implemented in user model", true)
+    assert(@standard.baseline_thermal_zone_demand_control_ventilation_required?(z))
+    assert(@standard.baseline_air_loop_hvac_demand_control_ventilation_required?(a))
+
+    refute_empty(a.airLoopHVACOutdoorAirSystem)
+    oa_sys = a.airLoopHVACOutdoorAirSystem.get
+    controller_oa = oa_sys.getControllerOutdoorAir
+    controller_mv = controller_oa.controllerMechanicalVentilation
+    refute(controller_mv.demandControlledVentilation)
+
+    climate_zone = 'ASHRAE 169-2013-5A'
+    @standard.model_set_baseline_demand_control_ventilation(@model, climate_zone)
+    assert(controller_mv.demandControlledVentilation)
+
+    # To produce the original crash, we need a second thermal zone on the
+    # AirLoopHVAC that does NOT need DCV
+    s2 = z.spaces.last
+    z2 = OpenStudio::Model::ThermalZone.new(@model)
+    s2.setThermalZone(z2)
+    a.addBranchForZone(z2)
+    # Assign a  DSOA that is not per floor area
+    dsoa = OpenStudio::Model::DesignSpecificationOutdoorAir.new(@model)
+    dsoa.setName("DSOA ACH")
+    dsoa.setOutdoorAirFlowAirChangesperHour(0.5)
+    @model.getSpaces.each {|s| s.setDesignSpecificationOutdoorAir(dsoa) }
+
+    assert(@standard.baseline_thermal_zone_demand_control_ventilation_required?(z))
+    refute(@standard.baseline_thermal_zone_demand_control_ventilation_required?(z2))
+    assert(@standard.baseline_air_loop_hvac_demand_control_ventilation_required?(a))
+    assert_equal(1, @model.getDesignSpecificationOutdoorAirs.size)
+    @standard.model_set_baseline_demand_control_ventilation(@model, climate_zone)
```
Add a test. It shows the issue originally

```
Minitest::UnexpectedError:         NameError: uninitialized constant OpenstudioStandards::ThermalZone
```

jmarrec commented on test/90_1_general/test_179d.rb:
```diff
@@ -511,4 +511,74 @@ def test_space_type_apply_internal_load_schedules_override
+    assert(@standard.baseline_air_loop_hvac_demand_control_ventilation_required?(a))
+    assert_equal(1, @model.getDesignSpecificationOutdoorAirs.size)
+    @standard.model_set_baseline_demand_control_ventilation(@model, climate_zone)
+    # One regular, one per area
+    assert_equal(2, @model.getDesignSpecificationOutdoorAirs.size)
+
+    refute_empty(s2.designSpecificationOutdoorAir)
+    refute_equal(dsoa, s2.designSpecificationOutdoorAir.get)
+    assert_equal("DSOA ACH to per-area", s2.designSpecificationOutdoorAir.get.nameString)
+    refute_equal(dsoa, s2.designSpecificationOutdoorAir.get)
+    assert_equal(0.0, s2.designSpecificationOutdoorAir.get.outdoorAirFlowAirChangesperHour)
+    assert_equal(0.0, s2.designSpecificationOutdoorAir.get.outdoorAirFlowperPerson)
+    assert_equal(0.0, s2.designSpecificationOutdoorAir.get.outdoorAirFlowRate)
+    assert(s2.designSpecificationOutdoorAir.get.outdoorAirFlowperFloorArea > 0)
+
+    z.spaces.each do |s|
+      assert_equal(dsoa, s.designSpecificationOutdoorAir.get)
+      assert_equal("DSOA ACH", s.designSpecificationOutdoorAir.get.nameString)
+      assert_equal(dsoa, s.designSpecificationOutdoorAir.get)
+      assert_equal(0.5, s.designSpecificationOutdoorAir.get.outdoorAirFlowAirChangesperHour)
+      assert_equal(0.0, s.designSpecificationOutdoorAir.get.outdoorAirFlowperFloorArea)
+      assert_equal(0.0, s.designSpecificationOutdoorAir.get.outdoorAirFlowperPerson)
+      assert_equal(0.0, s.designSpecificationOutdoorAir.get.outdoorAirFlowRate)
+    end
```
Complete testing

jmarrec commented on lib/openstudio-standards/standards/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.Model.rb:
```diff
@@ -1288,6 +1288,72 @@ def model_add_apxg_dcv_properties(model)
     # if a zone does not have this additional property, it means it was not served by airloop.
   end
 
+  # Convert total minimum OA requirement to a per-area value.
+  #
+  # @param thermal_zone [OpenStudio::Model::ThermalZone] OpenStudio ThermalZone object
+  # @return [Boolean] returns true if successful, false if not
+  def thermal_zone_convert_outdoor_air_to_per_area(thermal_zone)
```
The backported method

jmarrec commented on lib/openstudio-standards/standards/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.Model.rb:
```diff
@@ -1299,7 +1365,7 @@ def model_set_baseline_demand_control_ventilation(model, climate_zone)
         air_loop_hvac.thermalZones.sort.each do |zone|
           unless baseline_thermal_zone_demand_control_ventilation_required?(zone)
-            OpenstudioStandards::ThermalZone.thermal_zone_convert_outdoor_air_to_per_area(zone)
+            thermal_zone_convert_outdoor_air_to_per_area(zone)
```
And the fix

jmarrec commented:
@JanghyunJK what's going on with the merge and revert merge here?! You deleted my test, the addition of the rb files to lib/openstudio-standards.rb, etc.

IMHO, we just resert this branch to f2e6b65026267fcdb9bcbf1d0d9fe30c1129434c (a clean commit) and force push again

JanghyunJK commented:
> @JanghyunJK what's going on with the merge and revert merge here?! You deleted my test, the addition of the rb files to lib/openstudio-standards.rb, etc.

I just couldn't figure out how to update my local branch properly via github desktop. so I tried something with github desktop and felt like that was not the right way (so reverted). I ended up just deleting my local branch and then re-cloned. I wish I know how to do `IMHO, we just resert this branch to https://github.com/jmarrec/openstudio-standards/commit/f2e6b65026267fcdb9bcbf1d0d9fe30c1129434c (a clean commit) and force push again`....

JanghyunJK commented:
let me see what i can learn in 30 mins....

JanghyunJK commented:
hard reset?

![image](https://github.com/user-attachments/assets/90908a99-b016-4dd3-bd6c-428fccbf6f26)

JanghyunJK commented:
and probs to this [7d67196](https://github.com/NREL/openstudio-bem-to-surrogate-gem/commit/7d671962a7c8768cbc704617d2270d132f44e9d0) instead of https://github.com/jmarrec/openstudio-standards/commit/f2e6b65026267fcdb9bcbf1d0d9fe30c1129434c?

jmarrec commented:
[7d67196](https://github.com/NREL/openstudio-bem-to-surrogate-gem/commit/7d671962a7c8768cbc704617d2270d132f44e9d0) is from another repo. 

I just force pushed back to the f2e6b65026267fcdb9bcbf1d0d9fe30c1129434c

In your git client, right click on the commit you want and do "reset to this commit - hard". If that's nto an option, checkjout another branch, delete your local os-version-fix, then check it out again from the remote.

JanghyunJK commented:
spot checks (site EUI, kWh/sqft) before (top, our latest run) and after (bottom, smaller run with fewer samples) this change:

-----------------------------------------------
![image](https://github.com/user-attachments/assets/6c943e86-ee60-4605-996d-4b0dc13e8ee3)
![image](https://github.com/user-attachments/assets/0b23f571-c79c-493c-954e-5a71ac919586)
-----------------------------------------------
![image](https://github.com/user-attachments/assets/e705d3e8-5f48-47e4-a551-6866135651af)
![image](https://github.com/user-attachments/assets/62aae26f-9c20-49e0-b0b2-6886465c8cf7)
-----------------------------------------------
![image](https://github.com/user-attachments/assets/1ae7de31-88eb-41b7-aae6-daeef9631f2a)
![image](https://github.com/user-attachments/assets/c07a1874-c960-4766-abac-2fcb567bab55)
-----------------------------------------------
![image](https://github.com/user-attachments/assets/ba6061e8-cad3-4dcd-97c7-b17808b12149)
![image](https://github.com/user-attachments/assets/bfbeb5d7-72f2-49c9-b286-b6ac9dc858ea)
-----------------------------------------------
![image](https://github.com/user-attachments/assets/e44a0ea5-0690-41ca-b1f2-d858da087096)
![image](https://github.com/user-attachments/assets/a4b4ee1b-b549-4806-b723-4775e475525e)
-----------------------------------------------
![image](https://github.com/user-attachments/assets/5e7abb12-1cfe-4efc-acaa-bbd808564e65)
![image](https://github.com/user-attachments/assets/82869eec-a257-4b00-9872-07772e3da9e7)
-----------------------------------------------

JanghyunJK commented:
more spot checks:

-----------------------------------------------
![image](https://github.com/user-attachments/assets/a6664a48-ef2a-4d88-ac7b-bd314167136f)
![image](https://github.com/user-attachments/assets/bcb0cad2-0249-4b22-a487-a4471dae9771)
-----------------------------------------------
![image](https://github.com/user-attachments/assets/261923d5-1551-46d4-8b0c-e0dc30fdc06e)
![image](https://github.com/user-attachments/assets/ef369012-2f32-489a-a024-e91473a57b8f)
-----------------------------------------------
![image](https://github.com/user-attachments/assets/577b8464-92ce-4a61-8a2b-066487192b18)
![image](https://github.com/user-attachments/assets/62618703-7500-4d90-b924-009088258abe)
-----------------------------------------------
![image](https://github.com/user-attachments/assets/dae13e73-f9a7-424a-bb61-ce8a042dd06f)
![image](https://github.com/user-attachments/assets/a0d20456-3c04-476b-add4-5a66223a9632)
-----------------------------------------------
![image](https://github.com/user-attachments/assets/0b1177b8-386b-4458-8b7b-4d524b9c8d3b)
![image](https://github.com/user-attachments/assets/4511f0f0-4f5a-4348-888e-e9974d0ee575)
-----------------------------------------------
![image](https://github.com/user-attachments/assets/c8f71ea2-c2c0-4602-94b4-40bd4a406b31)
![image](https://github.com/user-attachments/assets/4ad169e7-aac1-4044-a695-2711d98f58ce)

JanghyunJK commented:
more on ordered savings (this plot is based on this PR):
![image](https://github.com/user-attachments/assets/5211da62-88fc-4f7a-954b-171a69876579)

JanghyunJK commented:
@brianlball I don't see any huge changes (👆) between before and after!
